### PR TITLE
Fixing rcd border zipper artefacts

### PR DESCRIPTION
**Pull request by jenshannoschwalm**
RCD has (as all demosaicers) a problem handling the border regions, the size of the border depends on the
demosaicer, for rcd theborder is 6 pixels wide. In other programs this is not as evident as in dt because
they crop after demosaicing and dt uses the roi concept.

So far rcd used the simple border approximation resulting in zipper artefacts.

I checked using other demosaicers to step in just for the border region in a better way, also such a
demosaicer should behave well performance wise. AMaZE misses a GPU implementation, VNG would be fine but
has some color drift, i tried the hq linear demosaicer (Malvar, HE,Cutler) but ended up in using PPG.
(Algo has been slightly modified to avoid calculating the area served by RCD)

There is a very subtle performance penalty, for CPU this is less than 1%, for GPU it seems to be less than 5%.
Also checked the integration suite showing no CPU vs GPU problems.

### CODE REVIEW

**Comment by jenshannoschwalm** [comment]
@Mark-64 this is probably what you we were asking for

File: src/iop/rcd_demosaic.c [hunk 1]
```diff
@@ -109,123 +109,190 @@ static INLINE float sqrf(float a)
...
+          const float m = fminf(pxm, pxM);
+          const float M = fmaxf(pxm, pxM);
+          color[1] = fmaxf(fminf(guessx * .25f, M), m);
         }
       }
-      approxit(out, cfa, sum, i * width + j, FCRCD(i, j));
+      else
+        color[1] = pc;
+
+      color[4] = 0.0f;
```
Comment (TurboGit) [inline src/iop/rcd_demosaic.c:211]: out of bound (declared float color[4])

File: src/iop/rcd_demosaic.c [hunk 2]
```diff
@@ -109,123 +109,190 @@ static INLINE float sqrf(float a)
...
-    float sum[6];
-    for(int j = border; j < width - border; j++)
-    {
-      for(int x = 0; x < 6; x++) { sum[x] = 0.0f; }
-      for(int i1 = i - 1; i1 < i + 2; i1++)
+      else
       {
-        for(int j1 = j - 1; j1 < j + 2; j1++)
+        // get 4-star-nbhood:
+        const float *ntl = buf - 4 - 4 * width;
```
Comment (TurboGit) [inline src/iop/rcd_demosaic.c:261]: I would add parentheses to clarify code:  `... - (4 * width)`

same below.

**Comment by jenshannoschwalm** [comment]
@TurboGit fixed both requests and slightly modified the order of GPU processing for lower mem pressure on the graphics card.

**Comment by Mark-64** [comment]
Tested.
It is way better than before, however PPG also had some small border degradation, and that is still barely visible, at least with test charts. The only algo which seems immune is Amaze.
I guess that will not be an issue with real life images, though.

**Comment by jenshannoschwalm** [comment]
PPG has a border interpolation of 3 pixels and does some minor correction for the inner two of those. So i think we are fine now :-)

**Comment by jenshannoschwalm** [comment]
As waiting for merge i added another fix for dual demosaicer mask calculation and display. Because of the border issues before this pr we can now be more relaxed about this.

@TurboGit i think ready to go

**Review by TurboGit — APPROVED** [review]
Thanks !

**Pull request merged**